### PR TITLE
Introduce hibernate envers, use it only for MdmLink, disable it for now, add to MdmLink Dao and unit test revisions.

### PR DESCRIPTION
**Pull request by lukedegruchy**
- Annotate MdmLink @Audited so that it will be versioned by Envers
- Support hibernate envers in the maven build, including the hdm.ddl plugin
- Override the defaults for the REVINFO table generation with more sensible names
- Ensure JpaConfig supports repositoryFactoryBeanClass = EnversRevisionRepositoryFactoryBean.class, which is required for hapi-fhir unit testing
- Add a new method to the MdmLink doas to retrieve the revision history and unit test this functionality
- Ensure that migration tasks create the envers tables in the same way as for a fresh install/new database

Closes https://github.com/hapifhir/hapi-fhir/issues/4632

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 1]
```diff
@@ -92,6 +92,92 @@ public HapiFhirJpaMigrationTasks(Set<String> theFlags) {
 		init620();
 		init640();
 		init660();
+		init670();
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:95]: I think you meant to add these to 660.  We are in 6.5-SNAPSHOT, which will release as 6.6.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java [hunk 2]
```diff
@@ -92,6 +92,92 @@ public HapiFhirJpaMigrationTasks(Set<String> theFlags) {
...
+		empiLink.addColumn("PARTITION_DATE")
+			.nullable()
+			.type(ColumnTypeEnum.DATE_ONLY);
+		empiLink.addColumn("PARTITION_ID")
+			.nullable()
+			.type(ColumnTypeEnum.INT);
+		empiLink.addColumn("GOLDEN_RESOURCE_PID")
+			.nonNullable()
+			.type(ColumnTypeEnum.LONG);
+		empiLink.addColumn( "TARGET_TYPE")
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:137]: huh: Weird. I know this matches the MPI_LINK definition, but HFJ_RESOURCE declares this varchar(100)

It is harmless to declare this as 100 to be safe.

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java [hunk 3]
```diff
@@ -90,6 +93,42 @@ public void testExpandPidsWorks() {
...
+		assertThat(mdmLink.getCreated(), is(nullValue()));
+		assertThat(mdmLink.getUpdated(), is(nullValue()));
+		myMdmLinkDaoSvc.save(mdmLink);
+		assertThat(mdmLink.getCreated(), is(notNullValue()));
+		assertThat(mdmLink.getUpdated(), is(notNullValue()));
+		assertTrue(mdmLink.getUpdated().getTime() - mdmLink.getCreated().getTime() < 1000);
+
+		final Revisions<Integer, MdmLink> mdmLinkHistoryCreate = myMdmLinkDaoSvc.findMdmLinkHistory(mdmLink);
+
+		mdmLinkHistoryCreate.forEach(revision -> System.out.println("CREATE revision = " + revision));
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java:108]: minor: use a logger at debug.

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java [hunk 4]
```diff
@@ -90,6 +93,42 @@ public void testExpandPidsWorks() {
...
+		assertThat(mdmLinkHistoryUpdate.stream().map(revision -> revision.getEntity().getMatchResult()).toList(),
+			equalTo(List.of(MdmMatchResultEnum.MATCH, MdmMatchResultEnum.NO_MATCH)));
+
+		myMdmLinkDaoSvc.deleteLink(mdmLink);
+
+		final Revisions<Integer, MdmLink> mdmLinkHistoryDelete = myMdmLinkDaoSvc.findMdmLinkHistory(mdmLink);
+		mdmLinkHistoryDelete.forEach(revision -> System.out.println("DELETE revision = " + revision));
+
+		assertThat(mdmLinkHistoryDelete.stream().map(Revision::getEntity).map(MdmLink::getMatchResult).toList(),
+			equalTo(listWithPossiblyNull(MdmMatchResultEnum.MATCH, MdmMatchResultEnum.NO_MATCH, null)));
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java:129]: minor: the `contains` matcher is gives better error messages than `equalTo` for lists.

File: hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java [hunk 5]
```diff
@@ -102,4 +141,8 @@ private MdmLink createPatientAndLinkTo(Long thePatientPid, MdmMatchResultEnum th
 		mdmLink.setSourcePersistenceId(runInTransaction(()->myIdHelperService.getPidOrNull(RequestPartitionId.allPartitions(), patient)));
 		return myMdmLinkDao.save(mdmLink);
 	}
+
+	private static <T> List<T> listWithPossiblyNull(T... items) {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-mdm/src/test/java/ca/uhn/fhir/jpa/mdm/dao/MdmLinkDaoSvcTest.java:145]: minor: I'm skeptical that this function is helping.  It doesn't save any characters, and is unfamiliar.  Inline.

File: pom.xml [hunk 6]
```diff
@@ -1770,6 +1770,16 @@
 					</exclusion>
 				</exclusions>
 			</dependency>
+			<dependency>
+				<groupId>org.hibernate</groupId>
+				<artifactId>hibernate-envers</artifactId>
+				<version>${hibernate_version}</version>
```
Comment (michaelabuckley) [inline pom.xml:1776]: Yeah. This is right.  Manage the versions here, and remove them above from the actual usage.

**Review by michaelabuckley — APPROVED** [review]
I assume that another PR will provide a config path - through StorageSettings or somewhere.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java (hunk 1 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/migrate/tasks/HapiFhirJpaMigrationTasks.java:95]: You're right.  I'll fix this.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/HapiFhirEnversRevision.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.SequenceGenerator;
+import javax.persistence.Table;
+import java.io.Serializable;
+
+@Entity
+@RevisionEntity
+@Table(name = "HFJ_REVINFO")
+public class HapiFhirEnversRevision implements Serializable {
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/HapiFhirEnversRevision.java:40]: That's correct.  It's just to override the names.   I'll javadoc this.

**Comment by lukedegruchy** [comment]
> I assume that another PR will provide a config path - through StorageSettings or somewhere.

Yes.  I'll do t his in another PR.

File: hapi-fhir-jpaserver-base/pom.xml [hunk 8]
```diff
@@ -476,6 +479,21 @@
...
+					<dependency>
+						<groupId>javax.xml.bind</groupId>
+						<artifactId>jaxb-api</artifactId>
+						<version>${jaxb_api_version}</version>
+					</dependency>
+					<dependency>
+						<groupId>com.sun.xml.bind</groupId>
+						<artifactId>jaxb-impl</artifactId>
+						<version>${jaxb_api_version}</version>
+					</dependency>
```
[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/pom.xml:496]: I didn't know that.  I'll fix this.

[the remark this answers is not in the document]
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-base/pom.xml]: Agreed that, ideally, child pom dependencies should not reference the versions and that the parent pom's dependencyManagement section should define these versions.

However, Maven reacts badly if I don't reference said versions in plugin dependencies, specifically.  Omitting said versions results in an ugly error:

```
[ERROR]   The project ca.uhn.hapi.fhir:hapi-fhir-jpaserver-base:6.5.4-SNAPSHOT (/Users/lukedegruchy/development/smilecdr/git/smilecdr-repos/slot5/hapi-fhir/hapi-fhir-jpaserver-base/pom.xml) has 1 error
[ERROR]     'build.plugins.plugin[de.jpdigital:hibernate56-ddl-maven-plugin].dependencies.dependency.version' for org.hibernate:hibernate-envers:jar is missing. @ line 482, column 18

```

**Pull request merged**